freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncStateApplier.java

```diff
@@ -89,7 +91,23 @@ public void applyStateDifferenceToIpa(String environmentCrn, FreeIpaClient freeI
 
         LOGGER.debug("Starting {} for {} group memberships ...", ADD_USERS_TO_GROUPS,
                 stateDifference.getGroupMembershipToAdd().size());
-        operations.addUsersToGroups(fmsToFreeipaBatchCallEnabled, freeIpaClient, stateDifference.getGroupMembershipToAdd(), warnings);
+        Set<String> largeGroupNames = umsUsersState.getGroupsExceedingThreshold();
+        if (largeGroupNames.isEmpty()) {
+            operations.addUsersToGroups(fmsToFreeipaBatchCallEnabled, freeIpaClient, stateDifference.getGroupMembershipToAdd(), warnings);
+        } else {
+            LOGGER.debug("group membership addition for large groups will not be batched: {}", largeGroupNames);
+            ImmutableMultimap.Builder<String, String> smallGroups = ImmutableMultimap.builder();
+            ImmutableMultimap.Builder<String, String> largeGroups = ImmutableMultimap.builder();
+            stateDifference.getGroupMembershipToAdd().asMap().forEach((groupName, users) -> {
+                if (largeGroupNames.contains(groupName)) {
+                    largeGroups.putAll(groupName, users);
+                } else {
+                    smallGroups.putAll(groupName, users);
+                }
+            });
+            operations.addUsersToGroups(fmsToFreeipaBatchCallEnabled, freeIpaClient, smallGroups.build(), warnings);
+            operations.addUsersToGroups(false, freeIpaClient, largeGroups.build(), warnings);
+        }
```

Comment:
I wouldn't add this code as it is here, at least move this to a separate method, but maybe into a class for easier testing
Also I would split the `else` part into 2 methods:
- addUsersToLargeGroups
- assUsersToSmallGroups
Instead of this for loop and multimap builders I would use stream to separate the 2:

```
    private void addGroupMembership(UmsUsersState umsUsersState, FreeIpaClient freeIpaClient, UsersStateDifference stateDifference, BiConsumer<String, String> warnings, boolean fmsToFreeipaBatchCallEnabled) throws FreeIpaClientException, TimeoutException {
        Set<String> largeGroupNames = umsUsersState.getGroupsExceedingThreshold();
        if (largeGroupNames.isEmpty()) {
            operations.addUsersToGroups(fmsToFreeipaBatchCallEnabled, freeIpaClient, stateDifference.getGroupMembershipToAdd(), warnings);
        } else {
            LOGGER.debug("group membership addition for large groups will not be batched: {}", largeGroupNames);
            addUsersToSmallGroups(freeIpaClient, stateDifference, warnings, fmsToFreeipaBatchCallEnabled, largeGroupNames);
            addUsersToLargeGroups(freeIpaClient, stateDifference, warnings, largeGroupNames);
        }
    }

    private void addUsersToLargeGroups(FreeIpaClient freeIpaClient, UsersStateDifference stateDifference, BiConsumer<String, String> warnings, Set<String> largeGroupNames) throws FreeIpaClientException, TimeoutException {
        HashMultimap<String, String> largeGroups = stateDifference.getGroupMembershipToAdd().entries().stream()
                .filter(entry -> largeGroupNames.contains(entry.getKey()))
                .collect(Multimaps.toMultimap(Map.Entry::getKey, Map.Entry::getValue, HashMultimap::create));
        operations.addUsersToGroups(false, freeIpaClient, largeGroups, warnings);
    }

    private void addUsersToSmallGroups(FreeIpaClient freeIpaClient, UsersStateDifference stateDifference, BiConsumer<String, String> warnings, boolean fmsToFreeipaBatchCallEnabled, Set<String> largeGroupNames) throws FreeIpaClientException, TimeoutException {
        HashMultimap<String, String> smallGroups = stateDifference.getGroupMembershipToAdd().entries().stream()
                .filter(entry -> !largeGroupNames.contains(entry.getKey()))
                .collect(Multimaps.toMultimap(Map.Entry::getKey, Map.Entry::getValue, HashMultimap::create));
        operations.addUsersToGroups(fmsToFreeipaBatchCallEnabled, freeIpaClient, smallGroups, warnings);
    }
```

Reply:
thanks for the suggestion. I pulled the code out to a separate class and it simplified the testing.

Reply:
thanks!